_posts/2018-09-07-redis-ruby-bloom-filter.md

```diff
@@ -0,0 +1,355 @@
+# [3992, 2262, 1788, 1970, 3185, 4135, 4957]
+```
+
+As a hashing function we use [CRC32](https://en.wikipedia.org/wiki/Cyclic_redundancy_check) with a custom seed per user that is the `billing_cycle_started_at` and a number of hashing functions based on the error rate (in this example we use an error rate of 1%).
+
+For the bloom filter to return consistent hashing indexes during a user's billing cycle it is important that all of the input parameters it depends on (`error_rate`, `@user.billing_cycle_started_at` and `@user.max_contacts`) does not change for the billing cycle until it will get reset. That is represented as a `fingerprint` that as we'll see later we'll use it as a redis key.
+
+
+### Redis bloom filter
+
+Redis supports `getbit` and `setbit` operations for [String](https://redis.io/commands#string) type that make the individual updates very simple. There is also a special data type for bloom filters called [rebloom](https://redislabs.com/blog/rebloom-bloom-filter-datatype-redis/) if you want to explore it.
+
+Here is our Redis bloom filter implementation:
+
+```ruby
+require 'redis'
+
+class RedisBloomFilter
+  MAX_TTL = 31 * 24 * 60 * 60 # max days in a month
+
+  def initialize(user)
+    @user = user
+  end
+
+  def insert(keys)
+    existing_indexes = redis.pipelined do
+      keys.each do |key|
+        bloom.indexes_for(key).map { |index| redis.setbit(filter_key, index, 1) }
+      end
+    end
+
+    new_keys_count = keys.length.times.count { |i|
+      existing_indexes[i * bloom.hash_functions, bloom.hash_functions].include?(0)
+    }
+
+    total = redis.incrby(counter_key, new_keys_count)
+
+    if total == new_keys_count
+      redis.expire(filter_key, MAX_TTL.to_i)
+      redis.expire(counter_key, MAX_TTL.to_i)
+    end
+  end
+
+  def count
+    redis.get(counter_key).to_i
+  end
+
+  def include?(key)
+    values = redis.pipelined do
+      bloom.indexes_for(key).map { |index| redis.getbit(filter_key, index) }
+    end
+
+    !values.include?(0)
+  end
+
+  def field
+    redis.get(filter_key)
+  end
+
+  private
+
+  def redis
+    @redis ||= Redis.new
+  end
+
+  def bloom
+    @bloom ||= BloomFilter.new(@user, error_rate: 1)
+  end
+
+  def filter_key
+    @filter_key ||= "bloom:filter:#{key_suffix}"
+  end
+
+  def counter_key
+    @counter_key ||= "bloom:counter:#{key_suffix}"
+  end
+
+  def key_suffix
+    @key_suffix ||= bloom.fingerprint
+  end
+end
+```
+The `RedisBloomFilter` uses the `BloomFilter` implementation to produce the indexes that it just stores/retrieves them to/from Redis. It also implements a counter of how many unique elements are added to the filter by increasing the count only when it detects a unique insert. It uses redis pipelined that sends operations in batch to avoid latency and improve performance by about 5-6 times. It also sets a TTLs on the keys to expire them after a month and it exposes the field value to be consumed from the in-memory filter.
```

Comment:
The first sentence isn't grammatically correct. Maybe removing the words `just` and `them` would fix it though

Reply:
Is `pipelined` a typo?

Reply:
Again, I'd be tempted to not reveal the exact TTLs time of a month for security purposes. Probably OK to mention that there's a TTL though